Update Sidebar UI

tlesicka opened the pull request:
---
category: Features
authors: [tlesicka]
---

Updating Sidebar.
Added collapsible Budget and Off Budget account groups.  Removed "Settings" from the "> More" expandible section.  Cleaned up "> More" to an up/down chevron.  Removed "Rename budget" from menu and added pencil icon next to budget name.  Changed scrollable area to just accounts.  Action Buttons and All accounts do not scroll.

Teprifer commented:
Thanks for putting this up, I've spotted a few minor things:

* Budget name icon positioning
Mouse not on side bar:
![image](https://github.com/user-attachments/assets/08933cce-8ea4-47a7-8a12-3072d09e71c0)
Mouse in side bar:
![image](https://github.com/user-attachments/assets/480be6c5-66f2-455c-a1fd-886b5477d5eb)
Mouse over budget name:
![image](https://github.com/user-attachments/assets/72c979ff-a4c6-4b5b-bbe5-902f84f60b11)

Notice how the drop down arrow moves to the left and is replaced by the edit name pencil when the mouse is over the budget name? This means a user could be aiming to click the drop down, but when the mouse gets there, it's no longer the drop down menu in that position and this will create a negative experience.
Could be a case of switching the pencil to the left side of the drop down icon, or move it to a menu item again, either/or.

* Top section:
1. There's extra vertical space between budget name and menu items
2. Menu items is showing less than before (schedules is hidden) - could add schedules back in by consuming the extra space above. This is mostly about keeping it consistent with the current behaviour and keeping them readily discoverable.
3. The more items has different show/hide behaviour,
   * In Edge if you're viewing one of the pages under More, when reloading the page the more menu stays expanded and collapses if a non-more item is being viewed.
   * In this PR the more menu is always collapsed after a page reload
   * Could remain with the current Edge approach, or remember expanded or collapsed state regardless of page viewed.


* Misc other thoughts...
   * de-prioritize closed accounts (less font weight/smaller size/other)
   * Adding 'more' back in and replace with 'less' when expanded, small font size so it doesn't take up more vertical than is already assigned for the toggle.
   * Bump 'all accounts' up a step font size to give it prominence over the below

Would also be good to get other opinions. :)

Jonathan-Fang commented:
What I like: 
- collapsible Budget and Off Budget account groups. Removed "Settings" from the "> More" expandible section.
- Removed "Rename budget" from menu and added pencil icon next to budget name. It's consistent with renaming accounts.

What I don't like: 
![image](https://github.com/user-attachments/assets/d9cdd411-cdc7-432e-99fa-7172344c104a)
![image](https://github.com/user-attachments/assets/387843e7-63d2-463a-90ad-5424fd1c9a0a)

As @Teprifer mentions, please re-add the More and Less as indicated, the chevron is kinda just lonely hanging out by itself.

![image](https://github.com/user-attachments/assets/45986a52-a0d9-4d75-9ec3-373b475c73bd)
A minor thing, would be great to minimize the whitespace here a little bit, it's just a little bit too big.

Great work! An improvement I didn't know I needed but now that I see it, I can't unsee it.

Jonathan-Fang commented:
@tlesicka I think you have addressed everything I brought up!

And minor thing, I'm wondering if we should capitalize "more" and "less" to fit with the rest of the app.

@Teprifer any thoughts on @tlesicka 's latest updates and the capitalizing "more" and "less"?

Teprifer commented:
I think my only feedback at this point is de-emphasising closed accounts.

@jfdoming have you had an opportunity to test it out?

tlesicka commented:
> I think my only feedback at this point is de-emphasising closed accounts.

Once I get this PR finished I'll start working on sortable account groups.  Account groups will be able to be renamed and sorted the same way that accounts are currently sorted.  This will remove the need to de-emphasis closed accounts.  Here's the discord discussion:
[Discord - Implement dividers for account list](https://discord.com/channels/937901803608096828/1280911541591347211/1286459037893787719)

tlesicka commented on packages/desktop-client/e2e/page-models/navigation.js:
```diff
@@ -43,11 +43,11 @@ export class Navigation {
+    // Expand the menu only if it is not already expanded
     if (!(await settingsLink.isVisible())) {
-      await this.page.getByRole('button', { name: 'More' }).click();
+      await this.page.getByRole('button', { name: 'Test Budget' }).click();
```
Moved Settings from expandable More section to the dropdown menu under the budget name.  Budget name for e2e is Test Budget, so this change from More to Test Budget was intentional.

jfdoming reviewed (changes requested):
Thanks for tackling this! Always appreciate fresh eyes on things like this. Some comments on my end:
- The "Closed" section should probably stay hidden until there are actually closed accounts
- The size of "More"/"Less" is now smaller than the rest of the menu items, which causes strange-looking jumping when the item is expanded. The "Less" item also jumps to the bottom, which is confusing if you don't know to look there. My preference would be that the "More"/"Less" item stays in the same location, and has the same size as the other elements for visual consistency.
- The text in account sidebar items appears to be selectable now, which was not the case previously. Mind fixing that?
- Something about the chevrons on the side is odd to me, both visually (different chevron than "More", alignment seems off) and UX-wise (clicking on account groups now feels like it should expand them, but it doesn't). On the latter, normally with Accordion components, clicking the "main" part of the component expands/collapses it. I think getting this right will be tricky and probably requires some design work. In the interest of making this PR easier to merge, my thinking is to revert the chevron behaviour for now and keep account groups always-expanded (as they currently are).

Thoughts?

tlesicka commented:
> * The "Closed" section should probably stay hidden until there are actually closed accounts

Updated.  Now, if any account group is empty, it will not be shown.

> * The size of "More"/"Less" is now smaller than the rest of the menu items, which causes strange-looking jumping when the item is expanded. The "Less" item also jumps to the bottom, which is confusing if you don't know to look there. My preference would be that the "More"/"Less" item stays in the same location, and has the same size as the other elements for visual consistency.

I think we'll disagree on this one.  I think keeping More in the middle disrupts the flow of the design.  I was trying for an accordion style where clicking the more expands to show more items.  If More stays in the middle, why not move those items to a menu for access?  Now, with more expanded, all action items are in the same location.  More itself is not an action item, so should not be the same size as the rest of the action items.

Does anyone else have any thoughts on the More/Less function?

> * The text in account sidebar items appears to be selectable now, which was not the case previously. Mind fixing that?

Fixed.  Previously, the text in the buttons could be selected. Now, no text on the sidebar can be selected.  Everything acts as a button.

> * Something about the chevrons on the side is odd to me, both visually (different chevron than "More", alignment seems off) and UX-wise (clicking on account groups now feels like it should expand them, but it doesn't). On the latter, normally with Accordion components, clicking the "main" part of the component expands/collapses it. I think getting this right will be tricky and probably requires some design work. In the interest of making this PR easier to merge, my thinking is to revert the chevron behaviour for now and keep account groups always-expanded (as they currently are).

Now that I look closely at the chevrons, yes, the alignment is off and I'll work on adjusting the margin/padding.  I agree that the account groups should act as accordion components.  If everyone agrees with removing the links from For budget/Off budget and making them more like accordion components, I will do that.  This change would make it where aligning the chevrons will be easier and I believe this would be a good change for Custom Account Groups too.  Personally, I would like the ability to expand/collapse account groups as I have around 15 For budget and another 15 Off budget accounts. Plus I don't look at For budget/Off budget transactions as a group, if I need to I'll use All accounts.  Not sure how many people find this {current} feature useful vs losing this feature for expandable/collapsible account groups.

Can I get more feedback on whether to remove the links to For budget/Off budget and make the account groups act as accordion components?

Teprifer commented:
> more/less
I like the general idea of when showing more, the menu items are un-interupted. I always felt it weird having more in the middle.

> Can I get more feedback on whether to remove the links to For budget/Off budget and make the account groups act as accordion components?

I'm not so sure about outright removing functionality, however, maybe (depending on feedback from others) it could be an option to move it elsewhere. Maybe a hover on for/off budget to have a -> widget to click instead? Dunno, just spit balling. 

But I think that's stepping outside this PR a bit? Since that'd have to be discussed a bit about how that'd best work.

Jonathan-Fang commented:
More/less function
- I'm fine with the way it is right now with the jumping. It's a bit abrupt but not sure what else other than like a quick animation where "more" is being pushed to the bottom and then becomes "less."
- Another idea is you could put More to the top, above Budget and below Test Budget. Then it wouldn't jump. That would be more accordion-style.
![image](https://github.com/user-attachments/assets/0531c1ad-51bf-44a5-8e89-6bfb35c0ca70)

Chevrons @tlesicka and @jfdoming 
- I disagree with removing the links from "for budget" and "off budget" as I am a user who utilizes these links.
- About making the account groups act as accordion components, right now the account groups are basically toggles; they work exactly the same as they would in an application like Notion. You click the toggle (chevron), and it expands by pointing downward. You click the toggle again, it goes back up, which collapses everything. If you want to make a UI change and make it like a darker background to make it look like a background was clicked, and the alignment can be exactly in the same place, I think it'd look better but it's not breaking the essentials of what this feature does.
![image](https://github.com/user-attachments/assets/850d718a-38b0-42d8-8543-26563788e587)
![image](https://github.com/user-attachments/assets/1d8335a1-c1cc-4bfa-bf00-288e8d03fce8)
![image](https://github.com/user-attachments/assets/cc691439-36c4-4170-92d6-18d474d90b8a)

tlesicka commented:
Thank you @Jonathan-Fang
> * I disagree with removing the links from "for budget" and "off budget" as I am a user who utilizes these links.

I am glad to know that people use these links.

> If you want to make a UI change and make it like a darker background to make it look like a background was clicked

I tried changing the background when it was clicked/expanded, but didn't really like it.  Instead the chevrons now have a hover highlight.

@jfdoming, the Action Buttons now have a sliding transition and isn't jumpy.

Jonathan-Fang commented:
@tlesicka I wanted to share that the Chevron hover highlight looks great!

The action button sliding animation looks great! I'm wondering if we should speed it up to match the snappiness of the rest of the application. Most buttons are click and instant change, so if we made it like the animation 2x-3x as fast, then maybe it would fit better.

What does everyone else think about the speed of the animation?

tlesicka commented on packages/desktop-client/src/components/sidebar/AccountGroup.tsx:
```diff
@@ -0,0 +1,100 @@
+// @ts-strict-ignore
+import React, { useState } from 'react';
```
Removed useState import

tlesicka commented on packages/desktop-client/src/components/sidebar/Sidebar.tsx:
```diff
@@ -162,29 +156,28 @@
+              Icon: SvgStoreFront,
+              to: '/payees',
+              hidable: true
+            },
```
Added the comma

tlesicka commented on packages/desktop-client/src/components/sidebar/AccountGroup.tsx:
```diff
@@ -0,0 +1,103 @@
+    borderRadius: '10px',
+    margin: '10px',
+    border: '2px solid var(--color-tableBorder)'
+  };
```
Commented out accountGroupStyle.  This style was used for dev work.

tlesicka commented on packages/loot-core/src/types/prefs.d.ts:
```diff
@@ -58,7 +58,8 @@
 export type LocalPrefs = MetadataPrefs &
   Partial<{
-    'ui.showClosedAccounts': boolean;
+    'ui.collapsedAccountGroups': Object;
```
Updated as indicated

tlesicka commented on packages/desktop-client/src/components/sidebar/Accounts.tsx:
```diff
@@ -1,168 +1,86 @@
-  onAddAccount: () => void;
-  onToggleClosedAccounts: () => void;
-  onReorder: OnDropCallback;
 };
```
Removed AccountsProps

tlesicka commented on packages/desktop-client/src/components/sidebar/BottomButtons.tsx:
```diff
@@ -0,0 +1,41 @@
+export function BottomButtons({
+  buttons,
+}: BottomButtonsProps) {
+  const { t } = useTranslation();
```
Removed t and useTranslate

tlesicka commented on packages/desktop-client/src/components/sidebar/AccountGroup.tsx:
```diff
@@ -0,0 +1,106 @@
+        to={groupTo}
+        query={groupQuery}
+        toggleAccounts={toggleAccounts}
+        collapsed={collapsed && collapsed[groupName.replace(/\s/g, "")]}
```
Updated as indicated

tlesicka commented on packages/desktop-client/src/components/sidebar/AccountGroup.tsx:
```diff
@@ -0,0 +1,106 @@
+    'ui.collapsedAccountGroups',
+  );
+
+  if (!collapsed || !collapsed.hasOwnProperty(groupName.replace(/\s/g, ""))) {
```
Updated as indicated

tlesicka reviewed (commented):
Fixed all coderabbitai issues

MatissJanis commented:
👋 Hey! I don't mean to dogpile on this PR, but would you mind splitting up the proposed changes in smaller, more isolated PRs? There are some suggestions that I think are not controversial and would be quickly approved and merged (i.e. the move of "+ add account" to the bottom, budget file edit icon), but others either need more polish, discussion or unfortunately would not be approved entirely. Splitting the proposal up into smaller chunks would allow this to move much more effectively.

tlesicka commented:
> 👋 Hey! I don't mean to dogpile on this PR, but would you mind splitting up the proposed changes in smaller, more isolated PRs? 

Not a problem.  Wasn't sure if I should make smaller but more numerous PRs or one PR that contains more changes.
Started another PR (#3554) to split this into multiple PRs.  Once PR #3554 is merged, I'll start on another aspect of this update.

> There are some suggestions that I think are not controversial and would be quickly approved and merged (i.e. the move of "+ add account" to the bottom, budget file edit icon), but others either need more polish, discussion or unfortunately would not be approved entirely.

@MatissJanis what things would not be approved?  Asking so that I don't spend time working on something that won't be approved.

MatissJanis commented:
> @MatissJanis what things would not be approved? Asking so that I don't spend time working on something that won't be approved.

These are the changes I noticed here (I might have missed some):

I have ranked them in order according to what I think will receive a lot of good feedback from the community (and thus be merged) to some that would require more fleshing out and/or would likely get a lot of resistance.

1. moving "+ add account" to the bottom
2. making the account list scrollable independently
3. adding edit icon besides budget file name
5. change of the "more/less" button 
4. chevrons to collapse account groups (IMO good idea, but visual execution will need more polish)
6. visual hierarchy for account headings & underlines
7. removal of separator (between accounts and main nav items) in sidenav
8. removal of "settings" from the side nav
9. "close budget file" move as the 1st item in the popover

RubenOlsen commented:
I am late to this party, and I have this very basic question: but what exactly are we trying to solve @tlesicka?

The original PR comment does not address this, so please explain this to me.

tlesicka commented:
> I am late to this party, and I have this very basic question: but what exactly are we trying to solve @tlesicka?
> 
> The original PR comment does not address this, so please explain this to me.

@RubenOlsen 
This PR was requested in a comment on issue #3414.  As stated in issue #3414:

> When a budget has a lot of accounts, the whole sidebar scrolls.

> I would suggest only the accounts should scroll and the "Budget", "Reports", and "Schedules" buttons should be fixed at the top. Additionally, the "+ Add Account" button should be fixed at the bottom. This separates the sidebar into four separate areas: Budget Name/Menu, Action Buttons, Account List, Add Account Button.

> Settings is doubled in both the menu and the "> More" action button and should be removed from one location.

> Remove "Rename budget" from the menu and instead put a pencil icon next to the name like on account view.

As I was working on these things, the scope of the whole PR increased and it was suggested to split this PR into smaller chunks.  There are aspects of the side bar that are doubled (settings in two locations), not in line with other aspects of AB (rename budget as menu item instead of pencil icon), or could have a better look/feel when using AB (scrolling and expandable more section).  This is what I was trying to solve.

This PR is now unused and will be closed soon in favor of smaller PR's with singular scope (such as #3554).

MatissJanis commented:
Just a very quick note: @tlesicka thank you for all the care and work you've put into this! It is truly APPRECIATED! 

As you can see: a lot of us are very opinionated about the side-nav. Which is a good problem to have, but at the same time it can be quite annoying for the person driving the change (_I just want to release my change! Stop sending me feedback! AAAhhhhh!!_). Please take our feedback comments as positive indicators that people **really care** about the work you're doing here!
